RCORE-2063 Fix some client resets potentially failing with AutoClientResetFailed if a new client reset condition occurred before the first one completed

michael-wb opened the pull request:
## What, How & Why?
If a new client reset occurs with a different action (e.g. rolled back to PBS), the original client reset tracking info (e.g. for migrated to FLX) will be removed and the new client reset will be allowed to continue. (Note: the action and error were originally included in this PR, but moved to a new PR #7649.

Added extra sync client hook events to capture different steps along the way during a client reset - this allows the "Test client migration and rollback with recovery" test to pause the client reset and roll back to PBS while the FLX migration client reset is in progress, effective reproducing the condition that was intermittently failing in the past. This test was previously taking 90+ secs to complete and about a third of this time was waiting for the reconnect timer to expire for the sync session that was active during migration and rollback. Added `handle_reconnect()` call after migration/rollback to cancel this timer, shaving around 30 secs off this test.

Fixes #7539

## ☑️ ToDos
* [X] 📝 Changelog update
* [X] 🚦 Tests (or not relevant)
* ~~[ ] C-API, if public C++ API changed~~
* ~~[ ] `bindgen/spec.yml`, if public C++ API changed~~

danieltabacaru commented:
Don't we want to fix the real issue too (clearing tracking the client reset in case of a rollback)?

michael-wb commented:
oh, yes - the test was originally rolling back during client reset on purpose to see what the issues are and the client reset tracking entry was being left over, causing the original failure.
I'll fix this, too.

michael-wb commented:
Added client reset error and action tracking to the client reset metadata storage (producing v2) and if the new client reset action is different from the current client reset action, then the older client reset tracking will be removed in favor of the new client reset. Also updated the "" test to reflect this operation:
```
Realm.Sync.Client.Reset - Connection[2]: Session[12]: Found a previous recoverable client reset of type: 'Recover' for 'MigrateToFLX' at: 2024-04-19 14:04:15.848902000
Realm.Sync.Client.Reset - Connection[2]: Session[12]: Originating client reset error: WrongSyncType: Client connected using partition-based sync when app has been migrated to flexible sync
Realm.Sync.Client.Reset - Connection[2]: Session[12]: New recoverable client reset of type: 'Recover' for 'RevertToPBS' is incompatible - clearing previous reset
Realm.Sync.Client.Reset - Connection[2]: Session[12]: New client reset error: WrongSyncType: Client connected using flexible sync when app has been reverted back to partition-based sync
```

jbreams commented:
Can we break this PR up into the three separate PRs? 1) adding the extra info to the client reset tracker to show what the original error was 2) changing it so that if a new client reset of a different type starts it will be allowed to continue 3) any changes to fix up the test ?

michael-wb commented:
> Can we break this PR up into the three separate PRs? 1) adding the extra info to the client reset tracker to show what the original error was 2) changing it so that if a new client reset of a different type starts it will be allowed to continue 3) any changes to fix up the test ?

Sure - I can do that

jbreams commented on CHANGELOG.md:
```diff
@@ -17,7 +17,7 @@
 
 ### Internals
-* None.
+* Fix client reset failure during sync migration due to previous incomplete client reset. ([PR #7542](https://github.com/realm/realm-core/pull/7542), since v13.11.0)
```
I think this actually belongs under the Fixed section rather than the Internals section.

jbreams replied:
Also would be helpful to describe how the client reset failure would manifest. Like what would the error message be for users?

michael-wb replied:
Sure

danieltabacaru reviewed (approved):
Looks good. Should we rename the PR to mention the bug being fixed (and not the test)?
